### PR TITLE
Import JSON Records from Kafka (from #24)

### PR DESCRIPTION
**Pull request by baunz**
This is a sketch on how to make this library import json records. The json conversion is rather basic as the main goal was to import the json as-is with the ``ÀS_JSON_DOC`` option, but the new ``RECORD_FIELDS`` option adds some value to the avro path as well: Right now a change in the record schema like field additions or removals or even a simple reordering of fields (which is possible) would break the import or just insert field values in the wrong columns.

### CODE REVIEW

**Comment by morazow** [comment]
Hello @baunz,

Thank a lot for the feedback and pull request!

Sorry for not replying earlier, I am going to review it soon.

File: doc/user_guide/user_guide.md [hunk 1]
```diff
@@ -34,11 +34,13 @@ Kafka topics with several partitions.
 
 ### Schema Registry
 
-Kafka connector requires Schema Registry. As a result, you should set it up
-together with a Kafka cluster.
+Kafka connector requires a Schema Registry when importing avro messages that are 
+serialized with a confluent schema registry on the producer side.
```
Comment (morazow) [inline doc/user_guide/user_guide.md:38]:
```suggestion
serialized with a Confluent schema registry on the producer side. 
```

File: doc/user_guide/user_guide.md [hunk 2]
```diff
@@ -263,15 +265,57 @@ Please notice that we convert Avro complex types to the JSON Strings. Use Exasol
 `VARCHAR(n)` column type to store them. Depending on the size of complex type,
 set the number of characters in the VARCHAR type accordingly.
 
-## Import From Kafka Cluster
+## Importing Raw JSON
+
+When specifying ``RECORD_FORMAT=JSON`` the connector expects a valid UTF-8
+serialized JSON record per message. 
+When using ``ÀS_JSON_DOC=true``, the record is inserted as a whole and
```
Comment (morazow) [inline doc/user_guide/user_guide.md:272]: Do you have unicode `A` on `AS_JSON_DOC` parameter? Could you please change it to regular `A`?

File: doc/user_guide/user_guide.md [hunk 3]
```diff
@@ -263,15 +265,57 @@ Please notice that we convert Avro complex types to the JSON Strings. Use Exasol
...
+# Import From Kafka Cluster
 
 Several property values are required to access the Kafka
 cluster when importing data from Kafka topics using the connector.
 
 You should provide these key-value parameters:
 
 - ``BOOTSTRAP_SERVERS``
-- ``SCHEMA_REGISTRY_URL``
+- ``SCHEMA_REGISTRY_URL`` (only when using avro)
```
Comment (morazow) [inline doc/user_guide/user_guide.md:318]: Let's remove it. It is already moved to optionals section.

File: doc/user_guide/user_guide.md [hunk 4]
```diff
@@ -405,6 +445,17 @@ configurations][kafka-consumer-configs].
 
 These are optional parameters with their default values.
 
+* ``SCHEMA_REGISTRY_URL`` - It specifies an URL to the Confluent [Schema
+  Registry][schema-registry] which stores Avro schemas as metadata. Schema
+  Registry will be used to parse the Kafka topic Avro data schemas.
+
+* ``RECORD_FORMAT`` - One of [avro, json]. Tje default value is **avro**
```
Comment (morazow) [inline doc/user_guide/user_guide.md:452]:
```suggestion
* ``RECORD_FORMAT`` - One of [avro, json]. The default value is **avro**.
```

File: src/main/scala/com/exasol/cloudetl/kafka/KafkaConsumerProperties.scala [hunk 5]
```diff
@@ -43,6 +43,15 @@ class KafkaConsumerProperties(private val properties: Map[String, String])
   final def getSingleColJson(): Boolean =
     isEnabled(AS_JSON_DOC)
 
+  /**
+   * Returns the type we expect on the kafka
+   * One of ['json', 'avro'] whereas 'avro is the default'
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/KafkaConsumerProperties.scala:48]:
```suggestion
   * Returns the type we expect on the Kafka.
   * It is one of ['json', 'avro'] whereas 'avro' is the default.
```

File: src/main/scala/com/exasol/cloudetl/kafka/KafkaConsumerProperties.scala [hunk 6]
```diff
@@ -43,6 +43,15 @@ class KafkaConsumerProperties(private val properties: Map[String, String])
...
+   */
+  final def getRecordFormat(): String = get(RECORD_FORMAT).getOrElse("avro")
+
+  final def getRecordFields(): Option[Seq[String]] =
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/KafkaConsumerProperties.scala:52]: Add short scaladoc.

File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/AvroDeserialization.scala [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+import com.exasol.cloudetl.kafka.{KafkaConnectorException, KafkaConsumerProperties}
+
+import io.confluent.kafka.serializers.{AbstractKafkaAvroSerDeConfig, KafkaAvroDeserializer}
+import org.apache.avro.generic.GenericRecord
+import org.apache.kafka.common.serialization.Deserializer
+import scala.jdk.CollectionConverters.MapHasAsJava
+
+/**
+ * Creates deserializers for avro records that are serialized with the confluent schema registry
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/deserialization/AvroDeserialization.scala:11]:
```suggestion
 * Creates deserializers for avro records that are serialized with the Confluent schema registry.
```

File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/JsonDeserialization.scala [hunk 8]
```diff
@@ -0,0 +1,37 @@
+package com.exasol.cloudetl.kafka.deserialization
+
+import com.exasol.cloudetl.kafka.{KafkaConnectorException, KafkaConsumerProperties}
+
+import org.apache.kafka.common.serialization.{Deserializer, StringDeserializer}
+
+/**
+ * Create deserializers for Json records
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/deserialization/JsonDeserialization.scala:8]:
```suggestion
 * Creates deserializers for JSON records.
```

File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/JsonDeserialization.scala [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+ */
+object JsonDeserialization extends RecordDeserialization {
+
+  override def getSingleColumnJsonDeserializer(
+    properties: KafkaConsumerProperties,
+    fields: Option[Seq[String]]
+  ): Deserializer[Seq[Any]] = {
+
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/deserialization/JsonDeserialization.scala:16]:
```suggestion
```

File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/RecordDeserialization.scala [hunk 10]
```diff
@@ -0,0 +1,28 @@
+package com.exasol.cloudetl.kafka.deserialization
+
+import com.exasol.cloudetl.kafka.KafkaConsumerProperties
+
+import org.apache.kafka.common.serialization.Deserializer
+
+/**
+ * A factory dor kafka deserializers
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/deserialization/RecordDeserialization.scala:8]:
```suggestion
 * A factory for Kafka deserializers.
```

File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/RecordDeserialization.scala [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+   * Create a deserializer that transforms the kafka record into a columnar structure.
+   * If a field list is provided, only those fields must be included in the seq
+   */
+  def getColumnDeserializer(
+    properties: KafkaConsumerProperties,
+    fields: Option[Seq[String]]
+  ): Deserializer[Seq[Any]]
+
+  /**
+   * Create a deserializer that emits the kafka record as a plain Json object
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/deserialization/RecordDeserialization.scala:22]:
```suggestion
   * Creates a deserializer that emits the Kafka record as a plain JSON object.
```

File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/RecordDeserialization.scala [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+import org.apache.kafka.common.serialization.Deserializer
+
+/**
+ * A factory dor kafka deserializers
+ */
+trait RecordDeserialization {
+
+  /**
+   * Create a deserializer that transforms the kafka record into a columnar structure.
+   * If a field list is provided, only those fields must be included in the seq
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/deserialization/RecordDeserialization.scala:14]:
```suggestion
   * Creates a deserializer that transforms the Kafka record into a columnar structure.
   * If a field list is provided, only those fields must be included in the sequence.
```

File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/ToStringDeserializer.scala [hunk 13]
```diff
@@ -0,0 +1,11 @@
+package com.exasol.cloudetl.kafka.deserialization
+
+import org.apache.kafka.common.serialization.Deserializer
+
+/**
+ * Emits the string representation record as single element seq
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/deserialization/ToStringDeserializer.scala:6]:
```suggestion
 * Emits the string representation record as single element sequence.
```

File: src/test/scala/com/exasol/cloudetl/kafka/deserialization/GenericRecordDeserializerTest.scala [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    record: GenericRecord,
+    fieldList: Option[Seq[String]]
+  ): Seq[Any] = {
+    val derse = mock[Deserializer[GenericRecord]]
+    when(derse.deserialize(ArgumentMatchers.anyString(), ArgumentMatchers.any[Array[Byte]]))
+      .thenReturn(record)
+    new GenericRecordDeserializer(fieldList, derse).deserialize("", Array.empty[Byte])
+  }
+
+  test("Must give all fields from record if field list is not provided") {
```
Comment (morazow) [inline src/test/scala/com/exasol/cloudetl/kafka/deserialization/GenericRecordDeserializerTest.scala:37]:
```suggestion
  test("must give all fields from record if field list is not provided") {
```

File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/JsonDeserializer.scala [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+              jsonNode =>
+                jsonNode.getNodeType match {
+                  case STRING  => jsonNode.asText()
+                  case NUMBER  => jsonNode.numberValue()
+                  case BOOLEAN => jsonNode.asBoolean()
+                  case _       => jsonNode.toString
+              }
+            )
+            .orNull
+      )
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/deserialization/JsonDeserializer.scala:33]: Extract inner mapping to a method.

File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/GenericRecordDeserializer.scala [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+import com.exasol.common.avro.AvroConverter
+
+import org.apache.avro.generic.GenericRecord
+import org.apache.kafka.common.serialization.Deserializer
+
+/**
+ * Extract a set of fields from an Avro GenericRecord. If no field list is defined,
+ * all fields in the record are emitted.
+ *
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/deserialization/GenericRecordDeserializer.scala:13]:
```suggestion
```

File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/GenericRecordDeserializer.scala [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+  private val converter = new AvroConverter()
+
+  final override def deserialize(topic: String, data: Array[Byte]): Seq[Any] = {
+    val record = deserializer.deserialize(topic, data)
+    val avroFields = record.getSchema.getFields.asScala
+    val fieldsToRead = fieldList
+      .map(_.map(filterField => Option(record.getSchema.getField(filterField))))
+      .getOrElse(avroFields.map(Option(_)))
+      .toSeq
+
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/deserialization/GenericRecordDeserializer.scala:29]:
```suggestion
```

File: src/main/scala/com/exasol/cloudetl/kafka/KafkaConsumerProperties.scala [hunk 18]
```diff
@@ -358,6 +369,21 @@ object KafkaConsumerProperties extends CommonProperties {
...
+   * Should be a comma separated list of fields present in the kafka record.
+   * When the source record is JSON this is required since the field order is not guaranteed
+   * in JSON record.
+   */
+  private[kafka] final val RECORD_FIELDS: String = "RECORD_FIELDS"
+
+  /**
+   * The serialization format of the topic we are reading.
+   * Either avro serialized with the confluent schema registry or json as plain string
+   * Needed to construct the correct deserializer.
```
Comment (morazow) [inline src/main/scala/com/exasol/cloudetl/kafka/KafkaConsumerProperties.scala:383]:
```suggestion
   * Either avro serialized with the Confluent schema registry or json as plain string
   * needed to construct the correct deserializer.
```

**Review by morazow — CHANGES_REQUESTED** [review]

**Comment by morazow** [comment]
Hello @baunz,

I have added some review suggestions. Please have a look.

Thanks again! It is nice feature.

**Comment by baunz** [comment]
Hi @morazow , thanks for taking the time too look at it! I will adress your comments in a new commit..

File: doc/user_guide/user_guide.md (hunk 3 above)
Comment (morazow) [reply doc/user_guide/user_guide.md:318]: Okay, yes that makes sense.

Let us rearrange, I have added suggestions.

File: doc/user_guide/user_guide.md [hunk 19]
```diff
@@ -263,15 +265,57 @@ Please notice that we convert Avro complex types to the JSON Strings. Use Exasol
...
+    -- Required for Kafka import UDF
+    KAFKA_PARTITION DECIMAL(18, 0),
+    KAFKA_OFFSET DECIMAL(36, 0),
+);
+```
+
+Note that the ``RECORD_FIELDS`` parameter is required when inserting JSON into
+columns as the order of fields in json records is not deterministic.
+
+# Import From Kafka Cluster
```
Comment (morazow) [inline doc/user_guide/user_guide.md:310]:
```suggestion
## Importing Avro Record
```

File: doc/user_guide/user_guide.md [hunk 20]
```diff
@@ -263,15 +265,57 @@ Please notice that we convert Avro complex types to the JSON Strings. Use Exasol
...
 
 Several property values are required to access the Kafka
 cluster when importing data from Kafka topics using the connector.
 
 You should provide these key-value parameters:
 
 - ``BOOTSTRAP_SERVERS``
-- ``SCHEMA_REGISTRY_URL``
+- ``SCHEMA_REGISTRY_URL`` (only when using avro)
```
Comment (morazow) [inline doc/user_guide/user_guide.md:318]:
```suggestion
- ``SCHEMA_REGISTRY_URL``
```

File: doc/user_guide/user_guide.md [hunk 21]
```diff
@@ -263,15 +265,57 @@ Please notice that we convert Avro complex types to the JSON Strings. Use Exasol
 `VARCHAR(n)` column type to store them. Depending on the size of complex type,
 set the number of characters in the VARCHAR type accordingly.
 
-## Import From Kafka Cluster
+## Importing Raw JSON
```
Comment (morazow) [inline doc/user_guide/user_guide.md:268]: Could you please move this section after the "Importing Avro Record" and before "Secure Connection"?

It is more easier in my eyes to read this way.

File: doc/user_guide/user_guide.md [hunk 22]
```diff
@@ -1,7 +1,7 @@
 # User Guide
 
 Exasol Kafka Connector Extension allows you to connect to Apache Kafka and
-import Avro formatted data from Kafka topics.  
+import Avro or Json formatted data from Kafka topics.  
 
 Using the connector you can import data from a Kafka topic into an Exasol table.
```
Comment (morazow) [inline doc/user_guide/user_guide.md:7]: Please update table of contents with new changes.

**Review by morazow — CHANGES_REQUESTED** [review]

**Review by morazow — CHANGES_REQUESTED** [review]
Added some suggestions for user guide.

File: doc/user_guide/user_guide.md (hunk 3 above)
Comment (baunz) [reply doc/user_guide/user_guide.md]: Thanks for the suggestions. I reordered the points, fixed the table of contents plus added an example for importing JSON similar to the avro section

**Review by morazow — APPROVED** [review]
Great! Thanks @baunz for the pull request!

**Comment by morazow** [comment]
I am going to merge this now. 

Release will come later, but meanwhile you could still use the assembled jar file.

**Pull request merged**